src/Utilities/DoubleDict.jl

```diff
@@ -0,0 +1,314 @@
+module DoubleDicts
+
+import MathOptInterface
+
+const MOI = MathOptInterface
+const CI{F,S} = MOI.ConstraintIndex{F,S}
+
+"""
+    DoubleDict
+
+Special dictionary to map ConstraintIndex to ConstraintIndex with minimal
+performance loss from the type instability due different constraint types.
+"""
+mutable struct DoubleDict{D, D2}
+    dict::D2{Tuple{DataType, DataType}, D{Int, Int}}
```

Comment:
We should maybe have `D <: AbstractDict{CI{F, S}, CI{F, S}}`. `getindex` would not be type-stable for `D2` but that would allow `D` to be used in a standalone way. When you are treating `F`-in-`S` constraints, you get the corresponding inner dict and then use this one in a standalone way to have a type-stable operations. `D` would even be a `DenseDict`.

Reply:
Thats similiar to what I tried to do with the `WithType`